### PR TITLE
[WIP][SQL] Refactor the implementation of `SparkDateTimeUtils#toJavaDate`

### PR DESCRIPTION
**Pull request by LuciferYang**

### CODE REVIEW

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 1]
```diff
@@ -223,12 +212,7 @@ trait SparkDateTimeUtils {
   def toJavaDate(days: Int): Date = {
     val rebasedDays = rebaseGregorianToJulianDays(days)
     val localMillis = Math.multiplyExact(rebasedDays, MILLIS_PER_DAY)
-    val timeZoneOffset = TimeZone.getDefault match {
-      case zoneInfo: TimeZone if zoneInfo.getClass.getName == zoneInfoClassName =>
-        getOffsetsByWallHandle.invoke(zoneInfo, localMillis, null).asInstanceOf[Int]
-      case timeZone: TimeZone =>
-        timeZone.getOffset(localMillis - timeZone.getRawOffset)
-    }
+    val timeZoneOffset = TimeZone.getDefault.getOffset(localMillis)
```
Comment (LuciferYang) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:215]: In https://github.com/apache/spark/pull/45716, I replaced the direct call to `sun.util.calendar.ZoneInfo` with the use of `MethodHandle` in order to make Spark 4.0 to be build with `-release`. 

However, referring to the `fromJavaDate` function, I wonder if we can directly use `TimeZone.getDefault.getOffset(localMillis)` here?


https://github.com/apache/spark/blob/03f4e45cd7e9e45149b45f3611efd85375e7a327/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala#L193-L198

All unit tests have passed with this pr. Is there any other reason that `sun.util.calendar.ZoneInfo` must be used here?

@hvanhovell @MaxGekk @dongjoon-hyun @HyukjinKwon

Comment (LuciferYang) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:215]: If this is feasible, I will update the PR description later

**Comment by nadavwr** [comment]
@LuciferYang If I may ask, why was this PR abandoned? I was hoping to be able to try out the 4.0 preview without --add-opens

**Comment by LuciferYang** [comment]
Under normal circumstances, there is no need to add additional `--add-opens` when using `4.0 preview`, because some `--add-opens` will be added by default during the job submission process.

If you are interested in this PR itself, you can try submitting it again :)

**Comment by nadavwr** [comment]
@LuciferYang I'm mostly looking for it to be solved for unit tests. I don't know why reflection would have been necessary in the first place here, so I don't see myself advocating for a PR to drop it -- but I definitely am hoping for it :)

**Pull request closed without merging**